### PR TITLE
fix #1973 Resolve a race that prevent decrement of REMAINING_TASK

### LINKED ISSUE #1973: BoundedElasticScheduler does not invoke any tasks after interruption.
I've stumbled upon a problem with BoundedElasticScheduler when there's high traffic creating many queued task on it. Invoking dispose() on scheduled tasks seems to break scheduler, to the point where it is not invoking any new tasks.

## Expected Behavior
Scheduler should correctly invoke tasks, after disposing all of its other tasks.

## Actual Behavior
Scheduler doesn't do anything after disposing all of its tasks, and scheduling a new one.

## Steps to Reproduce
In the following example, behavior is as follows:

* ~520 tasks are scheduled
* ~300 scheduled tasks are processed correctly
* When dispose() is called all of scheduled tasks are cancelled
* On most scheduled tasks InterruptedException occurs, except ~15 last ones (505-520)
* Scheduled task after dispose() call is never invoked.

```java
Logger logger = LoggerFactory.getLogger("Test");

    @Test
    public void test() throws InterruptedException {
        Scheduler scheduler =
                Schedulers.newBoundedElastic(30, Schedulers.DEFAULT_BOUNDED_ELASTIC_QUEUESIZE, "test-scheduler");
        Disposable subscription = Flux.range(0, 100000000)
                .flatMap(i -> {
                    logger.info("Scheduling {}", i);
                    return Mono.fromSupplier(() -> invokeLongOperation(i))
                            .subscribeOn(scheduler)
                            .publishOn(Schedulers.parallel())
                            .doOnError(throwable -> logger.error("Error occurred in thread {} ", i, throwable))
                            .doOnCancel(() -> logger.info("Thread {} cancelled", i));
                })
                .subscribe(i -> logger.info("Thread {} finished", i),
                           throwable -> logger.error("Error ", throwable));
        Thread.sleep(10000);
        logger.info("Invoking dispose");
        subscription.dispose();
        Thread.sleep(1000);
        Mono.fromSupplier(() -> invokeLongOperation(999999999))
                .subscribeOn(scheduler)
                .publishOn(Schedulers.parallel())
                .subscribe((i) -> logger.info("Consumed value {} ", i),
                           throwable -> logger.error("Exception"));
        logger.info("Sleeping");
        Thread.sleep(100000);
    }

    private int invokeLongOperation(Integer i) {
        logger.info("Invoking long operation {}", i);
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            logger.error("Thread {} interrupted", i);
        }
        return i;
    }
```

##Environment

* Reactor version used: 3.3.0
* JVM version: 1.8.0

### LINKED ISSUE #1992: Multiple subscribeOn calls with a BoundedElasticScheduler can cause deadlock
The following code, containing two `subscribeOn` calls never completes:

```java
final Scheduler scheduler = Schedulers.newBoundedElastic(1, Integer.MAX_VALUE, "test");
        
final Mono<Integer> integerMono = Mono.fromSupplier(() -> 1)
                .subscribeOn(scheduler)
                .subscribeOn(scheduler);

        integerMono.block();
```

The issue is unique to `BoundedElasticScheduler`. Other bounded schedulers do not show this behaviour. If one `subscribeOn` is removed, then it does run to completion.

## Expected Behavior
As per https://projectreactor.io/docs/core/release/reference/#_the_subscribeon_method, 

> Only the earliest subscribeOn call in the chain is actually taken into account. 

Therefore the code above should be equivalent to having a single `subscribeOn`. 

## Actual Behavior
The code never terminates. A stack dump shows it is awaiting a `CountDownLatch`

```java
"Test worker" #13 prio=5 os_prio=31 cpu=310.15ms elapsed=672.58s tid=0x00007faa55190800 nid=0x5b03 waiting on condition  [0x000070000a3dc000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.4/Native Method)
        - parking to wait for  <0x00000007ff86c888> (a java.util.concurrent.CountDownLatch$Sync)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.4/LockSupport.java:194)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.4/AbstractQueuedSynchronizer.java:885)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(java.base@11.0.4/AbstractQueuedSynchronizer.java:1039)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(java.base@11.0.4/AbstractQueuedSynchronizer.java:1345)
        at java.util.concurrent.CountDownLatch.await(java.base@11.0.4/CountDownLatch.java:232)
        at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:81)
        at reactor.core.publisher.Mono.block(Mono.java:1663)
        at test.SingleThreadedExecutorSubscribeOnTest.getOneFromMono(SingleThreadedExecutorSubscribeOnTest.java:39)
        at test.SingleThreadedExecutorSubscribeOnTest.subscribeOn_twice_for_single_threaded_boundedElastic_should_complete(SingleThreadedExecutorSubscribeOnTest.java:
```

## Steps to Reproduce
The following unit test illustrates two bounded schedulers which _do_ work as expected (`fromExecutor`, `fromSingle`), and the one which doesn't (`newBoundedElastic`)

```java
class SingleThreadedExecutorSubscribeOnTest {
    @Test
    void subscribeOn_twice_for_single_executor_should_complete() {
        getOneFromMono(Schedulers.newSingle("test"));
    }

    @Test
    void subscribeOn_twice_for_single_threaded_boundedElastic_should_complete() {
        // This never completes
        getOneFromMono(Schedulers.newBoundedElastic(1, Integer.MAX_VALUE, "test"));
    }

    @Test
    void subscribeOn_twice_for_single_threaded_executor_should_complete() {
        final Scheduler executorScheduler = Schedulers.fromExecutor(
                new ThreadPoolExecutor(0, 1,
                        60L, TimeUnit.SECONDS,
                        new LinkedBlockingDeque<>()));

        getOneFromMono(executorScheduler);
    }

    private void getOneFromMono(final Scheduler scheduler) {
        final Mono<Integer> integerMono = Mono.fromSupplier(() -> 1)
                .subscribeOn(scheduler)
                .subscribeOn(scheduler);

        integerMono.block();
    }
}
```

## Possible Solution
Not researched in detail, however it appears that having _n_ `subscribeOn` calls in a chain subscribing with a scheduler with _n-1_ `threadCap` triggers the issue. 

## Your Environment
* Reactor version(s) used: io.projectreactor:reactor-core:3.3.1.RELEASE
* JVM version (`javar -version`):
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)

* OS and version (eg `uname -a`): MacOS

### PR DESCRIPTION
**Pull request by simonbasle**
~we might not need a complete reimplementation as in #2040 after all, if this fixes #1973 and #1992 (to be verified for the later)~

this fixes #1973 only

### CODE REVIEW

File: reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java [hunk 1]
```diff
@@ -680,6 +681,47 @@ public void defaultBoundedElasticConfigurationIsConsistentWithJavadoc() {
 				.isEqualTo(Schedulers.DEFAULT_BOUNDED_ELASTIC_QUEUESIZE);
 	}
 
+	@Test
+	public void raceActiveWorkerDisposeAndDeferredDirectDispose() {
+		BoundedElasticScheduler boundedElasticScheduler = new BoundedElasticScheduler(
+				1, 2, r -> new Thread(r,"raceActiveDeferredDirectDispose"), 3);
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java]:
```suggestion
				1, 2, r -> new Thread(r, "raceActiveDeferredDirectDispose"), 3);
```

File: reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java [hunk 2]
```diff
@@ -680,6 +681,47 @@ public void defaultBoundedElasticConfigurationIsConsistentWithJavadoc() {
...
+		int maxRounds = 1_000_000;
+		AtomicInteger incrementWon = new AtomicInteger();
+		AtomicInteger rejected = new AtomicInteger();
+		int rounds = 0;
+		while(rounds++ < maxRounds) {
+			try {
+				if (rounds % 10 == 0) { //from time to time reverse the race
+					RaceTestUtils.race(
+							activeWorker::dispose,
+							boundedElasticScheduler.schedule(incrementWon::incrementAndGet)::dispose
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java]: schedule always happens before the race starts due to the method ref on `dispose`. Is it intentionally?

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java]: good catch, that was at first glance not that bad but in fact we'd better wrap it in a `Runnable`. updated!

File: reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java [hunk 3]
```diff
@@ -680,6 +681,47 @@ public void defaultBoundedElasticConfigurationIsConsistentWithJavadoc() {
...
+			try {
+				if (rounds % 10 == 0) { //from time to time reverse the race
+					RaceTestUtils.race(
+							activeWorker::dispose,
+							() -> boundedElasticScheduler.schedule(incrementWon::incrementAndGet).dispose()
+					);
+				}
+				else {
+					RaceTestUtils.race(
+							boundedElasticScheduler.schedule(incrementWon::incrementAndGet)::dispose,
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java]: Missing lambda too

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java]: 🤦‍♂ fixing...

**Pull request merged**